### DISCUSSION TITLE
How to implement a key-based discriminated union

### DISCUSSION THREAD

**Discussion by ghost** [pydantic/pydantic#4180] [Question] [answered]
Hello everyone!

I have a JSON file that looks like:

```json
{
  "name": "pydantic",
  "pet": {
    "cat": {
      "name": "Happy",
      "color": "white",
      "cat_property": "1"
    }
  }
}
```

or looks like:

```json
{
  "name": "pydantic",
  "pet": {
    "dog": {
      "name": "Happy",
      "color": "white",
      "dog_property": "1"
    }
  }
}
```

It looks very similar to what [a discriminated union](https://pydantic-docs.helpmanual.io/usage/types/#discriminated-unions-aka-tagged-unions) can offer. However, `Field(discriminator="...")` works only on fields of the submodel. 

So even if I write something like:

```python3

class Cat(NamedTuple):
    type_: Literal["cat"]
    name: str
    color: str
    cat_property: str

class Dog(NamedTuple):
    type_: Literal["dog"]
    name: str
    color: str
    dog_property: str

class Pet(BaseModel):
    type_: Literal["cat", "dog"]
    ...
```

It is nowhere near what I want.

Unfortunately, I have no way to change the schema of these JSON files to add `"type": "dog"` inside the submodel. Does `pydantic` currently offer any possible way to deal with this? Thank you guys!

**Comment by samuelcolvin** [comment, marked as answer]
Currently I don't know of an easy way.

It'll be possible in pydantic v2, we're just discussing the feature on https://github.com/samuelcolvin/pydantic-core/issues/102

**Reply by ghost** [reply]
I have tried the below:

```python3
class Cat(NamedTuple):
    type_: Literal["cat"]
    options: Dict

class Dog(NamedTuple):
    type_: Literal["dog"]
    options: Dict

class Person(BaseModel):
    pet: Union[Cat, Dog] = Field(discriminator="type_")
    ...
```

Got the error:

```
  Discriminator 'type_' is missing in value (type=value_error.discriminated_union.missing_discriminator; discriminator_key=type_)
```

The `type_` property is already in the submodel. Perhaps this error occurs because discrimination runs before the value of `type_` was loaded. Do we have any workaround for this instead? Thank you!

**Reply by ghost** [reply]
Or are there any ways I can use a validator to transform it into something discriminated union can understand?

**Reply by dhurley-arine** [reply]
I recently had a similar issue trying to create a model for very similarly structured JSON and stumbled upon this thread while looking for solutions. While I also initially wanted to solve this using Unions like you, I ultimately figured out that the best way to solve it was using a [typed dict](https://docs.pydantic.dev/usage/types/#typeddict), as below:

```python
class Cat(NamedTuple):
    name: str
    color: str
    cat_property: str


class Dog(NamedTuple):
    name: str
    color: str
    dog_property: str


# `total=False` means keys are non-required
class Pet(TypedDict, total=False):
    cat: Cat
    dog: Dog


class Person(BaseModel):
    name: str
    pet: Pet
    ...
```

**Reply by dmontagu** [reply]
@cliff-lau-cloverhealth I wasn't able to get this working quickly using `NamedTuple`, but you can get this working if `Cat` and `Dog` are instead BaseModels by using a root_validator to inject the discriminator field:
```python
from typing import Literal, Dict, Union

from pydantic import BaseModel, Field, root_validator


class Cat(BaseModel):
    type_: Literal["cat"]
    options: Dict


class Dog(BaseModel):
    type_: Literal["dog"]
    options: Dict


PET_TYPE_KEYS = {'cat', 'dog'}


class Person(BaseModel):
    pet: Union[Cat, Dog] = Field(discriminator="type_")

    @root_validator(pre=True)
    def insert_type(cls, values):
        pet = values.get('pet', {})
        type_ = None
        for key in pet:
            if key in PET_TYPE_KEYS:
                type_ = key
                break
        if type_ is not None:
            values['pet'] = {'type_': type_, 'options': pet}
        return values


print(
    Person.parse_obj({
        "name": "pydantic",
        "pet": {
            "cat": {
                "name": "Happy",
                "color": "white",
                "cat_property": "1"
            }
        }
    })
)
# pet=Cat(type_='cat', options={'cat': {'name': 'Happy', 'color': 'white', 'cat_property': '1'}})
```
